boards: arm: add toradex verdin imx8m plus board

GabrielHAFs opened the pull request:
Add Verdin iMX8M Plus board with i.MX8MP SoC and ARM Cortex-M7 processor. Port and documentation are based on the NXP MIMX8MM EVK board. This code is intended to be used with the Cortex-M7.

Commit:
boards: arm: add toradex verdin imx8m plus board

Add Verdin iMX8M Plus board with i.MX8MP SoC and ARM Cortex-M7 processor.
Add two targets (DDR and ITCM) for the iMX8M Plus board.
Port and documentation are based on NXP MIMX8MM EVK board.
This code is intented to be used with the Cortex-M7.

Signed-off-by: Gabriel Freitas <gabriel.freitas@toradex.com>

danieldegrasse commented on boards/arm/verdin_imx8mp_m7/verdin_imx8mp_m7_defconfig:
```diff
@@ -0,0 +1,20 @@
+CONFIG_PINCTRL=y
+
+CONFIG_CODE_ITCM=y
+CONFIG_CODE_DDR=n
```
Is `CONFIG_CODE_DDR=n` needed here? This symbol is a Kconfig choice.

GabrielHAFs replied:
I'm not sure if this is the right approach, but I added both configs `CONFIG_CODE_ITCM` (enabled by default) and `CONFIG_CODE_DDR` (disabled by default) as options, as explained in the `index.rst`. If the user chooses to use the ITCM RAM, leave it as is. If it decides to use DDR RAM, enable the `CONFIG_CODE_DDR=y` and CONFIG_CODE_ITCM=n`.

GabrielHAFs replied:
Changed to two different targets, one for DDR and the other for TCM, as explained on https://github.com/zephyrproject-rtos/zephyr/pull/61713#discussion_r1376573982

danieldegrasse reviewed (commented):
Hi @GabrielHAFs, thank you for the contribution. Could you please address the compliance failures?

danieldegrasse commented on boards/arm/verdin_imx8mp_m7/verdin_imx8mp_m7.dts:
```diff
@@ -0,0 +1,54 @@
+		zephyr,flash = &itcm;
+		zephyr,sram = &dtcm;
+		/* DDR */
+		/* zephyr,flash = &ddr_code; */
```
Something you could consider to enable running from DDR or TCM is to create two different board targets, this is something that has been done for some of the IMX boards: https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/arm/mimx8mp_evk

GabrielHAFs replied:
Thanks for the suggestion @danieldegrasse. I have changed the code to add two targets, such as NXP.

mmahadevan108 commented:
cc @Mani-Sadhasivam

MaureenHelm commented:
@GabrielHAFs can you please have a look at the review comments?

Commit:
dts: add support for uart1 usage on imx8ml_m7 devicetree include file

Add support for UART1 usage by adding uart1 node and configuration
to the i.MX 8ML devicetree include.

Signed-off-by: Gabriel Freitas <gabriel.freitas@toradex.com>

GabrielHAFs commented on dts/arm/nxp/nxp_imx8ml_m7.dtsi:
```diff
@@ -167,6 +167,14 @@
 			status = "disabled";
 		};
 
+		uart1: uart@30860000 {
```
This is resolved on the commit 0ffcb49d73a08ca31c202dad49d1c116f5b91222.

GabrielHAFs commented on boards/arm/verdin_imx8mp_m7/verdin_imx8mp_m7.dts:
```diff
@@ -0,0 +1,54 @@
+};
+
+&uart1 {
+	status = "disabled";
```
This was resolved. Thanks for the comment.

GabrielHAFs commented on boards/arm/verdin_imx8mp_m7/verdin_imx8mp_m7.dts:
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2017,2019 NXP
```
Resolved on the last commit.

GabrielHAFs commented on boards/arm/verdin_imx8mp_m7/verdin_imx8mp_m7.yaml:
```diff
@@ -0,0 +1,25 @@
+#
+# Copyright (c) 2017, NXP
```
Resolved on the last commit.

danieldegrasse commented:
Thanks for the updates- a few things:
- Please make sure the commit enabling the use of UART1 is the first commit in the series, before the one enabling the board itself
- Please take a look at the CI failures, it looks like the board is failing to build a few tests.

GabrielHAFs commented:
@danieldegrasse I have addressed the issues and changed the order of the commits, as requested.

danieldegrasse commented on boards/arm/verdin_imx8mp_m7/verdin_imx8mp_m7_ddr.yaml:
```diff
@@ -0,0 +1,22 @@
+  - gnuarmemb
+  - xtools
+testing:
+  ignore_tags:
```
Question here- do we need `ignore_tags` set in order to pass CI on this board? Otherwise, I'd like to avoid defining it.

GabrielHAFs replied:
I tested locally and it seems to not be necessary. I'll remove the `ignore_tags`.

danieldegrasse commented on boards/arm/verdin_imx8mp_m7/verdin_imx8mp_m7_itcm_defconfig:
```diff
@@ -0,0 +1,17 @@
+CONFIG_CLOCK_CONTROL=y
+CONFIG_UART_CONSOLE=y
+CONFIG_SERIAL=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
```
Is `CONFIG_UART_INTERRUPT_DRIVEN` needed as part of the minimal set of Kconfigs to enable this platform? If not, I would like to drop it here

GabrielHAFs replied:
It's not necessary. I am dropping it.

danieldegrasse commented on boards/arm/verdin_imx8mp_m7/verdin_imx8mp_m7_ddr_defconfig:
```diff
@@ -0,0 +1,17 @@
+CONFIG_CLOCK_CONTROL=y
+CONFIG_UART_CONSOLE=y
+CONFIG_SERIAL=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
```
Same question here about `CONFIG_UART_INTERRUPT_DRIVEN`

GabrielHAFs replied:
Same as https://github.com/zephyrproject-rtos/zephyr/pull/61713#discussion_r1377930090

kartben commented on boards/arm/verdin_imx8mp_m7/doc/index.rst:
```diff
@@ -0,0 +1,290 @@
+.. _verdin_imx8mp_m7:
+
+Toradex Verdin iMX8M Plus SoM
+
```
FWIW, this improperly formatted heading is causing the page to not show up in the documentation. This, plus the fact that the "build hello world" instructions were basically broken (wrong board name) makes me think that the documentation for this page could have use a more thorough review (at a minimum testing the page properly shows up in the documentation ;-))

Just spent 30 minutes cleaning things up in #65240, please have a look so that we can get the table of contents for Arm boards back to a better state :)

![image](https://github.com/zephyrproject-rtos/zephyr/assets/128251/510ab18a-75df-4636-bb31-d75ab029450e)

danieldegrasse replied:
Thanks for pointing this out, and sorry for missing it initially.

@GabrielHAFs can you take a look at this?

kartben replied:
I obviously linked to the wrong PR -- #65420
